File: java/space-age/src/main/java/SpaceAge.java
class SpaceAge {

    private final int EARTH_SECONDS_PER_YEAR = 31557600;
    private final double EARTH   = 1.0;
    private final double JUPITER = 11.862615;
    private final double MARS    = 1.8808158;
    private final double MERCURY = 0.2408467;
    private final double NEPTUNE = 164.79132;
    private final double SATURN  = 29.447498;
    private final double URANUS  = 84.016846;
    private final double VENUS   = 0.61519726;

    private double ageInSeconds;

    SpaceAge(double seconds) {
        this.ageInSeconds = seconds;
    }

    double getSeconds() {
        throw new UnsupportedOperationException("Delete this statement and write your own implementation.");
    }

    private double onPlanet(double planetOrbitalPeriodRelativeToEarth) {
        double secondsOnPlanet = EARTH_SECONDS_PER_YEAR * planetOrbitalPeriodRelativeToEarth;
        return ageInSeconds / secondsOnPlanet;
    }

    double onEarth() {
        return onPlanet(EARTH);
    }

    double onMercury() {
        return onPlanet(MERCURY);
    }

    double onVenus() {
        return onPlanet(VENUS);
    }

    double onMars() {
        return onPlanet(MARS);
    }

    double onJupiter() {
        return onPlanet(JUPITER);
    }

    double onSaturn() {
        return onPlanet(SATURN);
    }

    double onUranus() {
        return onPlanet(URANUS);
    }

    double onNeptune() {
        return onPlanet(NEPTUNE);
    }

}

File: ruby/space-age/space_age.rb
class SpaceAge
  EARTH_YEAR_IN_SECONDS = 60 * 60 * 24 * 365.25
  def initialize(seconds)
    @seconds = seconds
  end

  YEAR_DURATION_RELATIVE_TO_EARTH = {
    mercury: 0.2408467,
    venus:     0.61519726,
    earth:     1.0,
    mars:      1.8808158,
    jupiter:  11.862615,
    saturn:   29.447498,
    uranus:   84.016846,
    neptune: 164.79132
  }.each do |planet, earth_years|
    define_method :"on_#{planet}" do
      @seconds.fdiv(EARTH_YEAR_IN_SECONDS * earth_years).round(2)
    end
  end
end